### ISSUE TITLE
Bug: Padding tokens are not masked in distillation losses

### ISSUE THREAD

**Issue by peremartra** [peremartra/optipfair#34]
<p>The distillation module (<code>optipfair.distillation</code>) computes all four loss components (task, logits/Skew KLD, trajectory, derivative) over <strong>all token positions including padding tokens</strong>. This dilutes the training signal proportionally to the amount of padding in each batch, biases the relative weights of loss components away from their intended meaning, and can substantially affect final model quality when training data has variable-length sequences padded to a common length.</p>
<p>I encountered this issue while preparing the experiments for Chapter 6 of <em>Rearchitecting LLMs</em> (the book this library supports). After fixing it in my notebooks and re-running the experiments at three different dataset sizes (2K / 15K / 40K samples), I confirmed the bug is real, reproducible, and meaningful in magnitude. The same code pattern is present in <code>optipfair.distillation</code>, so the bug exists in the library too.</p>
<h2>Root cause</h2>
<p>In <code>optipfair/distillation/trainer.py</code>, inside the training loop:</p>
<pre><code class="language-python">labels = inputs["input_ids"].clone()
</code></pre>
<p>Labels are constructed as a verbatim copy of <code>input_ids</code>, with no masking of padding token positions. As a result:</p>
<ol>
<li>The cross-entropy <code>ignore_index=-100</code> in <code>compute_distillation_loss</code> never fires, because <code>labels</code> never contains <code>-100</code>.</li>
<li>There is no shared signal that downstream loss components could use to identify padding.</li>
</ol>
<p>In <code>optipfair/distillation/loss.py</code>, none of the four loss components apply masking:</p>
<ul>
<li><strong>Task loss</strong>: relies on <code>ignore_index=-100</code>, which is never present → padding tokens contribute.</li>
<li><strong>Skew KLD (<code>loss_logits</code>)</strong>: uses <code>.sum(dim=-1).mean()</code> over all positions → padding included.</li>
<li><strong>Trajectory loss (<code>loss_trajectory</code>)</strong>: cosine similarity averaged over all positions → padding included.</li>
<li><strong>FDD derivative loss (<code>loss_derivative</code>)</strong>: cosine similarity of layer-to-layer deltas averaged over all positions → padding included.</li>
</ul>
<h2>Why this matters</h2>
<p>Concretely, with <code>padding="max_length"</code> and variable-length sequences (a very common setup), a significant fraction of every batch is padding tokens. In my experiments with <code>max_length=512</code> on the Cosmopedia dataset, this fraction was substantial.</p>
<p>Symptoms of the bug:</p>
<ol>
<li><strong>Diluted signal</strong>: the effective learning rate on real tokens is lower than what <code>alpha</code>, <code>beta</code>, <code>gamma</code>, <code>delta</code> suggest. The weights don't mean what users think they mean.</li>
<li><strong>Trivial-pattern learning</strong>: the student is asked to predict <code>pad_token → pad_token</code> (often <code>eos → eos</code> when <code>pad_token = eos_token</code>), which is easy and inflates loss magnitudes spuriously without contributing useful capability.</li>
<li><strong>FDD specifically suffers</strong>: derivative losses on padding tokens are essentially noise (the "rate of change" between layers over padding has no semantic meaning), so the FDD term gets regularized toward matching teacher noise on padding positions.</li>
</ol>
<h2>Empirical evidence</h2>
<p>I ran the same three-experiment comparison (Last-Layer-Logits-only, Selected-Layer-Features, Selected-Layer-Advanced-KD with Skew KLD + FDD) on Gemma-3-270M with 4 layers removed, at 2K / 15K / 40K samples, both with the buggy code and with masking properly applied.</p>
<p><strong>At 40K samples (the final book experiment):</strong></p>

Model | PPL (buggy) | PPL (fixed) | PPL retention (buggy) | PPL retention (fixed)
-- | -- | -- | -- | --
Last Logits | 12.1 | 11.9 | 106.6% | 108.4%
Sel Feat (cosine) | 12.1 | 11.9 | 106.6% | 108.1%
Sel Adv (Skew KLD + FDD) | 11.5 | 11.3 | 112.0% | 113.9%

<p>Two things to note. First, all methods improve by ~4 PPL with the fix, which is substantial (~17% reduction). Second, the apparent advantage of Advanced KD (Skew KLD + FDD) over plain feature alignment <strong>disappears entirely</strong> when masking is applied. With the bug, Adv looked clearly better than Feat (21.3 vs 24.0). With the fix, they're indistinguishable (19.7 vs 19.6).</p>
<p>This is the most important consequence of the bug: at small-to-medium data regimes, <strong>it can make distillation strategies look meaningfully different from each other when they actually aren't</strong>. Users running ablations to choose between strategies might draw incorrect conclusions.</p>
<h2>Proposed solution</h2>
<p>The fix has two parts: mask labels in the trainer, and propagate the mask through all four loss components.</p>
<h3>Part 1: <code>trainer.py</code></h3>
<p>Replace:</p>
<pre><code class="language-python">labels = inputs["input_ids"].clone()
model_inputs = {k: v for k, v in inputs.items() if k != "labels"}
</code></pre>
<p>With:</p>
<pre><code class="language-python">labels = inputs["input_ids"].clone()
if "attention_mask" in inputs:
    labels[inputs["attention_mask"] == 0] = -100
model_inputs = {k: v for k, v in inputs.items() if k != "labels"}
</code></pre>
<p>This is the minimal, backward-compatible change. When <code>attention_mask</code> is present (the standard case), padding gets the conventional <code>-100</code> marker. When it isn't, behavior is unchanged.</p>
<p>A slightly more ambitious version: if the user already passes <code>labels</code> in <code>inputs</code>, respect them and only auto-generate when absent. This gives advanced users a way to use custom label masking (e.g., for instruction-tuning style targets where prompt tokens should also be ignored).</p>
<p><strong>My recommendation</strong>: implement both — respect user-provided <code>labels</code> if present, otherwise auto-generate from <code>input_ids</code> and <code>attention_mask</code>. The full snippet would be:</p>
<pre><code class="language-python">if "labels" in inputs:
    labels = inputs["labels"].clone()
else:
    labels = inputs["input_ids"].clone()
    if "attention_mask" in inputs:
        labels[inputs["attention_mask"] == 0] = -100
model_inputs = {k: v for k, v in inputs.items() if k != "labels"}
</code></pre>
<p>This covers (a) the common case where users only pass <code>input_ids</code> + <code>attention_mask</code>, and (b) the advanced case where users want custom label masking for instruction tuning.</p>
<h3>Part 2: <code>loss.py</code></h3>
<p>Derive a shared mask from labels and apply it to all four components. Below is the full replacement function:</p>
<pre><code class="language-python">def compute_distillation_loss(
    student_logits: torch.Tensor,
    teacher_logits: torch.Tensor,
    student_hiddens: Optional[List[torch.Tensor]],
    teacher_hiddens: Optional[List[torch.Tensor]],
    labels: torch.Tensor,
    layer_map: Optional[List[int]] = None,
    alpha: float = 0.6,
    beta: float = 0.4,
    gamma: float = 0.0,
    delta: float = 0.0,
    temperature: float = 2.0,
    skew_alpha: float = 0.4,
) -&gt; Tuple[torch.Tensor, Dict[str, float]]:
    """
    Compute the compound knowledge distillation loss.

    All four loss components operate only on valid (non-padding) tokens,
    identified by labels != -100. Callers are responsible for setting
    labels to -100 at padding positions before calling this function.
    """
    device = student_logits.device

    if (
        (gamma &gt; 0 or delta &gt; 0)
        and layer_map is None
        and student_hiddens is not None
        and teacher_hiddens is not None
    ):
        layer_map = create_layer_map_uniform(len(student_hiddens), len(teacher_hiddens))

    # 1. Task loss (already respects ignore_index=-100)
    shift_logits = student_logits[..., :-1, :].contiguous()
    shift_labels = labels[..., 1:].contiguous()
    loss_task = F.cross_entropy(
        shift_logits.view(-1, shift_logits.size(-1)),
        shift_labels.view(-1),
        ignore_index=-100,
    )

    # Shared valid-token mask for KL, trajectory, and derivative losses
    valid_mask = (shift_labels != -100)              # [batch, seq_len - 1]
    num_valid = valid_mask.sum().clamp(min=1)

    # 2. Skew KLD (logits alignment) — masked
    with torch.no_grad():
        student_probs = F.softmax(student_logits[..., :-1, :] / temperature, dim=-1)
        teacher_probs = F.softmax(teacher_logits[..., :-1, :] / temperature, dim=-1)
        mixed_probs = skew_alpha * student_probs + (1 - skew_alpha) * teacher_probs

    student_log_probs = F.log_softmax(student_logits[..., :-1, :] / temperature, dim=-1)
    kl_elementwise = student_probs * (student_log_probs - torch.log(mixed_probs + 1e-9))
    kl_per_token = kl_elementwise.sum(dim=-1)        # [batch, seq_len - 1]
    loss_logits = (kl_per_token * valid_mask).sum() / num_valid
    loss_logits = loss_logits * (temperature ** 2)

    # 3. Trajectory loss (hidden state cosine similarity) — masked
    if gamma &gt; 0 and student_hiddens is not None and teacher_hiddens is not None:
        loss_trajectory = 0.0
        flat_mask = valid_mask.reshape(-1)

        for student_idx, teacher_idx in enumerate(layer_map):
            student_h = student_hiddens[student_idx][:, :-1, :].contiguous()
            teacher_h = teacher_hiddens[teacher_idx][:, :-1, :].contiguous()

            student_flat = student_h.reshape(-1, student_h.size(-1))
            teacher_flat = teacher_h.reshape(-1, teacher_h.size(-1))

            student_norm = F.normalize(student_flat, p=2, dim=1)
            teacher_norm = F.normalize(teacher_flat, p=2, dim=1)

            cos_sim_per_token = (student_norm * teacher_norm).sum(dim=1)
            cos_sim = (cos_sim_per_token * flat_mask).sum() / num_valid
            loss_trajectory += 1 - cos_sim

        loss_trajectory = loss_trajectory / len(layer_map)
    else:
        loss_trajectory = torch.tensor(0.0, device=device)

    # 4. FDD derivative loss — masked
    loss_derivative = torch.tensor(0.0, device=device)
    if delta &gt; 0 and student_hiddens is not None and teacher_hiddens is not None:
        num_derivatives = 0
        flat_mask = valid_mask.reshape(-1)

        for student_idx in range(len(layer_map) - 1):
            teacher_idx = layer_map[student_idx]
            teacher_idx_next = layer_map[student_idx + 1]

            student_delta = (student_hiddens[student_idx + 1][:, :-1, :]
                             - student_hiddens[student_idx][:, :-1, :])
            teacher_delta = (teacher_hiddens[teacher_idx_next][:, :-1, :]
                             - teacher_hiddens[teacher_idx][:, :-1, :])

            student_delta_flat = student_delta.reshape(-1, student_delta.size(-1))
            teacher_delta_flat = teacher_delta.reshape(-1, teacher_delta.size(-1))

            student_delta_norm = F.normalize(student_delta_flat, p=2, dim=1)
            teacher_delta_norm = F.normalize(teacher_delta_flat, p=2, dim=1)

            cos_sim_deriv_per_token = (student_delta_norm * teacher_delta_norm).sum(dim=1)
            cos_sim_deriv = (cos_sim_deriv_per_token * flat_mask).sum() / num_valid
            loss_derivative += 1 - cos_sim_deriv
            num_derivatives += 1

        if num_derivatives &gt; 0:
            loss_derivative = loss_derivative / num_derivatives

    total_loss = (
        alpha * loss_task
        + beta * loss_logits
        + gamma * loss_trajectory
        + delta * loss_derivative
    )

    loss_dict = {
        "total": total_loss.item(),
        "task": loss_task.item(),
        "logits": loss_logits.item(),
        "trajectory": loss_trajectory.item(),
        "derivative": loss_derivative.item(),
    }

    return total_loss, loss_dict
</code></pre>
<p>Key changes vs. the current implementation:</p>
<ol>
<li><code>valid_mask</code> and <code>num_valid</code> derived from <code>shift_labels != -100</code> once, reused by KL/trajectory/derivative.</li>
<li>KL formula itself is unchanged (still the Hinton-style <code>student_probs * (student_log_probs - log(mixed_probs))</code> from DistiLLM-2); only the reduction changes from <code>.mean()</code> over all positions to a masked sum-divided-by-valid-tokens.</li>
<li>Trajectory and derivative shift hidden states with <code>[:, :-1, :]</code> to align with the <code>seq_len - 1</code> mask, then apply <code>flat_mask</code> before averaging.</li>
<li>Function docstring updated to document the <code>-100</code> convention for the caller.</li>
</ol>
<h2>Backward compatibility note</h2>
<p>This change will alter the absolute magnitudes of all loss components (they will appear larger because they're no longer diluted by easy padding positions). However, in my experiments at three different scales, <strong>the relative ordering between distillation strategies is preserved</strong>, and all final model metrics (PPL, benchmark accuracy) either improve or stay within ~1 pp of the buggy version.</p>
<p>Users with hard-coded loss-magnitude expectations in logging or early-stopping logic may need to retune. Users tracking final model quality should see equal or slightly better results.</p>
<h2>Documentation impact</h2>
<p>The semantic meaning of <code>alpha</code>, <code>beta</code>, <code>gamma</code>, <code>delta</code> changes slightly with the fix: they now control the weighting of loss components evaluated <em>on real tokens only</em>, rather than averaged over all positions (including padding). Any documentation or README sections describing these weights should remain accurate in spirit, but if there are recommended weight presets, those may benefit from re-tuning. From my experiments, the existing defaults still produce sensible results after the fix; no retuning is strictly required.</p>
<h2>Suggested testing</h2>
<p>A targeted regression test:</p>
<ol>
<li>Construct a small batch with two sequences: one full-length (no padding) and one heavily padded.</li>
<li>Call <code>compute_distillation_loss</code> with all four weights &gt; 0 and <code>labels</code> properly masked at padding positions.</li>
<li>Vary the amount of padding in the second sequence (e.g., 25%, 50%, 75% padding) while keeping the non-padded content identical.</li>
<li>Assert that the four loss values remain stable across padding levels (within a small numerical tolerance, e.g., 1e-4 relative).</li>
</ol>
<p>With the bug, varying padding amount changes losses significantly. With the fix, padding is invisible to the loss — which is the invariant we want.</p>
<h2>Offer to contribute</h2>
<p>Happy to open a PR with the changes above plus a regression test if the team agrees with the diagnosis and proposed fix. Let me know if you'd prefer a different approach on any of the design choices (e.g., the labels handling logic) and I'll align with that.</p></body></html>
